lib/profileCache.js

```diff
@@ -11,41 +13,41 @@ const batch = require('./batch');
 // sets of scopes, but we have to limit it to a small set of known
 // possibilities in order to efficiently implement the "drop"
 // method below.
-const CACHEABLE_SCOPES = {
-  // Trusted reliers all use the main "profile" scope.
-  'profile': true,
-  // Untrusted reliers tend to request these three scopes.
-  'display_name+email+uid': true
-};
 
+// Trusted reliers all use the main "profile" scope.
+const SCOPE_PROFILE = ScopeSet.fromArray(['profile']);
+// Untrusted reliers tend to request these three scopes.
+const SCOPE_UNTRUSTED = ScopeSet.fromArray(['profile:display_name', 'profile:email', 'profile:uid']);
+
+const CACHEABLE_SCOPES = [
+  SCOPE_PROFILE.getScopeValues(),
+  SCOPE_UNTRUSTED.getScopeValues()
+];
 
-// Extract the profile-related scopes for a request.
-// This removes non-profile scopes and returns either
-// the string 'profile' or a sorted plus-separated list
-// of subscopes like 'display_name+email'.
 
-function extractProfileScopes(req) {
+function getProfileCacheKey(req) {
   const creds = req.auth.credentials;
-  // We cache the results to avoid doing this loop multiple times.
-  if (creds.hasOwnProperty('_fxaCachedProfileScopes')) {
-    return creds._fxaCachedProfileScopes;
-  }
-  const subScopes = {};
-  for (let i = 0, len = creds.scope.length; i < len; i++) {
-    let scope = creds.scope[i];
-    if (scope.endsWith(':write')) {
-      scope = scope.substr(0, scope.length - ':write'.length);
+  // We stash the result on the request object to avoid doing these checks multiple times.
+  if (! creds.hasOwnProperty('_fxaProfileCacheKey')) {
+    // By default, requests are not cached.
+    creds._fxaProfileCacheKey = null;
+    const uid = creds.user;
+    const scope = ScopeSet.fromArray(creds.scope);
+    // If they have read access to the full profile, we can cache it.
+    // We use the bare uid as the key for b/w compat with previous
+    // versions of this code, to avoid starting over with an empty cache.
+    if (scope.contains(SCOPE_PROFILE)) {
+      creds._fxaProfileCacheKey = uid;
     }
-    if (scope === 'profile') {
-      creds._fxaCachedProfileScopes = 'profile';
-      return 'profile';
-    } else if (scope.startsWith('profile:')) {
-      subScopes[scope.substr('profile:'.length)] = true;
-    } else if (scope === 'email') {
-      subScopes.email = true;
+    // If they're using a set of scopes common to untrusted reliers,
+    // cache under a key for that specific set of scopes.  It's important
+    // that they don't have other profile-related scopes, otherwise we
+    // give them cached data that's missing some fields they're entitled to.
+    else if (scope.contains(SCOPE_UNTRUSTED) && ! scope.difference(SCOPE_UNTRUSTED).intersects(SCOPE_PROFILE)) {
+      creds._fxaProfileCacheKey = `${uid}:scoped:display_name+email+uid`;
```

Comment:
does `scoped` have special significance? What happens if someone requests e.g., `profile:display_name profile:email`, will that be cached under the same key? I'm having a hard time understanding fully 
`scope.difference(SCOPE_UNTRUSTED).intersects(SCOPE_PROFILE)`

Reply:
> does scoped have special significance? 
No special significance, it's just a string to help clarify what's going on in the cache key name.
>  What happens if someone requests e.g., profile:display_name profile:email, will that be cached under the same key?

No, because the cached item contains the uid, which they don't have permission to view.

I'll see if I can add some more explanation in comments.